src/hotspot/share/code/nmethod.cpp

```diff
@@ -1414,92 +1258,63 @@ void nmethod::unlink_from_method() {
       return false;
     }
 
-    // The caller can be calling the method statically or through an inline
-    // cache call.
-    if (!is_osr_method() && !is_not_entrant()) {
+    if (is_osr_method()) {
+      // This logic is equivalent to the logic below for patching the
+      // verified entry point of regular methods.
+      // this effectively makes the osr nmethod not entrant
+      invalidate_osr_method();
+    } else {
+      // The caller can be calling the method statically or through an inline
+      // cache call.
       NativeJump::patch_verified_entry(entry_point(), verified_entry_point(),
-                  SharedRuntime::get_handle_wrong_method_stub());
+                                       SharedRuntime::get_handle_wrong_method_stub());
     }
 
-    if (is_in_use() && update_recompile_counts()) {
-      // It's a true state change, so mark the method as decompiled.
-      // Do it only for transition from alive.
+    if (update_recompile_counts()) {
+      // Mark the method as decompiled.
       inc_decompile_count();
     }
 
-    // If the state is becoming a zombie, signal to unregister the nmethod with
-    // the heap.
-    // This nmethod may have already been unloaded during a full GC.
-    if ((state == zombie) && !is_unloaded()) {
-      nmethod_needs_unregister = true;
-    }
-
-    // Must happen before state change. Otherwise we have a race condition in
-    // nmethod::can_convert_to_zombie(). I.e., a method can immediately
-    // transition its state from 'not_entrant' to 'zombie' without having to wait
-    // for stack scanning.
-    if (state == not_entrant) {
-      mark_as_seen_on_stack();
-      OrderAccess::storestore(); // _stack_traversal_mark and _state
+    BarrierSetNMethod* bs_nm = BarrierSet::barrier_set()->barrier_set_nmethod();
+    if (bs_nm == nullptr || !bs_nm->supports_entry_barrier(this)) {
+      // If nmethod entry barriers are not supported, we won't mark
+      // nmethods as on-stack when they become on-stack. So we
+      // degrade to a less accurate flushing strategy, for now.
+      mark_as_maybe_on_stack();
     }
 
     // Change state
-    if (!try_transition(state)) {
-      // If the transition fails, it is due to another thread making the nmethod more
-      // dead. In particular, one thread might be making the nmethod unloaded concurrently.
-      // If so, having patched in the jump in the verified entry unnecessarily is fine.
-      // The nmethod is no longer possible to call by Java threads.
-      // Incrementing the decompile count is also fine as the caller of make_not_entrant()
-      // had a valid reason to deoptimize the nmethod.
-      // Marking the nmethod as seen on stack also has no effect, as the nmethod is now
-      // !is_alive(), and the seen on stack value is only used to convert not_entrant
-      // nmethods to zombie in can_convert_to_zombie().
-      return false;
-    }
+    bool success = try_transition(not_entrant);
```

Comment:
I suppose in a further RFE, you could rename try_transition or do the transition inline.  Does it still CAS?

Reply:
It no longer does CAS because we always told the CompiledMethod_lock now when fiddling with this transition. The transition from not_installed to in_use can still fail, if the nmethod became not_entrant before it became in_use. I think the not_installed state can be removed altogether, but I think that's a separate RFE.